icu4c/source/common/unicode/uset.h

```diff
@@ -1285,4 +1325,408 @@ U_CAPI UBool U_EXPORT2
+    }
+
+    USetRangeIterator end() const {
+        return USetRangeIterator(uset, rangeCount, rangeCount);
+    }
+
+    // TODO: cbegin() & cend() ??
+
+private:
+    const USet *uset;
+    int32_t rangeCount;
+};
+
+/**
+ * TODO API docs
+ * @draft ICU 76
+ */
+class USetStringIterator {
+public:
+    USetStringIterator(const USetStringIterator &other) = default;
+
+    bool operator==(const USetStringIterator &other) const {
+        // NOTE: I think we don't need to compare count given private constructor
+        // and assuming we don't compare iterators across the set being modified.
+        // We might even skip comparing uset.
+        // Unless we want operator==() to be "correct" for more than iteration.
+        return uset == other.uset && index == other.index;
+    }
+
+    bool operator!=(const USetStringIterator &other) const { return !operator==(other); }
+
+    std::u16string_view operator*() const {
+        if (index < count) {
+            int32_t length;
+            const UChar *uchars = uset_getString(uset, index, &length);
+            // assert uchars != nullptr;
+            return { ConstChar16Ptr(uchars), (uint32_t)length };
+        }
+        return {};
+    }
+
+    USetStringIterator &operator++() {  // pre-increment
+        ++index;
+        return *this;
+    }
+
+    USetStringIterator operator++(int) {  // post-increment
+        USetStringIterator result(*this);
+        ++index;
+        return result;
+    }
+
+private:
+    friend class USetStrings;
+
+    USetStringIterator(const USet *uset, int32_t index, int32_t count)
+            : uset(uset), index(index), count(count) {}
+
+    const USet *uset;
+    int32_t index;
+    int32_t count;
+};
+
+/**
+ * TODO API docs
+ * @draft ICU 76
+ */
+class USetStrings {
+public:
+    USetStrings(const USet *uset) : uset(uset), count(uset_getStringCount(uset)) {}
+
+    USetStrings(const USetStrings &other) = default;
+
+    USetStringIterator begin() const {
+        return USetStringIterator(uset, 0, count);
+    }
+
+    USetStringIterator end() const {
+        return USetStringIterator(uset, count, count);
+    }
+
+    // TODO: cbegin() & cend() ??
+
+private:
+    const USet *uset;
+    int32_t count;
+};
+
+/**
+ * TODO API docs
+ * @draft ICU 76
+ */
+class USetElementIterator {
+public:
+    USetElementIterator(const USetElementIterator &other) = default;
+
+    bool operator==(const USetElementIterator &other) const {
+        // NOTE: I think we don't need to compare rangeCount & end given private constructor
+        // and assuming we don't compare iterators across the set being modified.
+        // We might even skip comparing uset.
+        // Unless we want operator==() to be "correct" for more than iteration.
+        return uset == other.uset && c == other.c && index == other.index;
+    }
+
+    bool operator!=(const USetElementIterator &other) const { return !operator==(other); }
+
+    UnicodeString operator*() const {
```

Comment:
Hm, when we discussed this one in ICU-TC, I was thinking of std::whateverstring, but of course we probably want UnicodeString here.
Does UnicodeString have the small-string optimization though?

Reply:
> Hm, when we discussed this one in ICU-TC, I was thinking of std::whateverstring, but of course we probably want UnicodeString here.

I chose to return a UnicodeString because it's easy to make one from a code point, and fits well with other ICU APIs.

> Does UnicodeString have the small-string optimization though?

yes